MessageBox: Extract in-line message links from the message content and show them as footlinks

preetmishra opened the pull request:
The initial commits extract in-line message links from the message content and add them below it to improve the reading experience. 

The commits are small and incremental in order to make reviewing manageable. Though, they could be squashed before merging.

The last commit adds in-line message links to `MsgInfoView` (needs discussion; see in-line comment).

Fixes #618 and #452 partially.

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -676,6 +673,7 @@ def transform_content(self) -> Tuple[None, Any]:
             padding_char = '░'
             self.indent_quoted_content(soup, padding_char)
 
+        self.link_index = 1
```
For maintaining a link counter, I couldn't make a recursive approach work and eventually had to go with an attribute. I could use pointers on how to make it clearer.

neiljp replied:
Could we use `state` for this?

preetmishra replied:
(Resolved in PM.)

preetmishra commented on zulipterminal/config/keys.py:
```diff
@@ -99,6 +99,11 @@
         'key_category': 'msg_actions',
     }),
+    ('MSG_LINKS', {
+        'keys': {'m'},
```
Which key would we want to warrant for `MSG_LINKS`?

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -476,14 +476,11 @@ def soup2markup(self, soup: Any, **state: Any) -> List[Any]:
-                        ('msg_link', '[' + text + ']' + '(' + link + ')'))
+                text = text if text else link
+                markup.append(
+                    ('msg_link', '{} [{}]'.format(text, self.link_index))
```
Currently, I am showing link as text whenever text is not present. This, however, could be noisy when the link is comparatively long.

neiljp replied:
Having duplicate content seems confusing and noisy, yes.

kaustubh-nair replied:
Yeah, I'd prefer only having one when there's no text.

preetmishra replied:
As per the recent discussion, the inclination is towards only using the link in either case.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -901,9 +901,12 @@ def __init__(self, controller: Any, title: str, *, zt_version: str,
-
+        for index, (text, link) in enumerate(msg['links'].items(), start=1):
+            widgets.append(
+                urwid.Text('[{}]: {} - {}'.format(index, text, link))
```
As link can be extremely long, I am inclining towards only showing text with mark - `[1]: Text` in `MsgLinksView`. Thoughts?

preetmishra replied:
Also, I haven't used `PopUpView` here as I am planning to use buttons from Shreya Malviya's old PR later (for handling internal/external links).

neiljp replied:
I'm not sure precisely what you mean here, but perhaps this warrants discussion in light of checkboxes being used with @kaustubh-nair 's stream settings work, to work towards a general solution?

kaustubh-nair replied:
Since both PRs are being worked on simultaneously, I think it would be better if we merge them both first and then do a refactor - this would save time as well as make it easier to ensure that the refactor works for both the PRs

preetmishra replied:
In the latest update, I have used the existing code from `PopUpView` but, yes, the use-case will likely evolve if we decide to incorporate new changes over the course of time. I believe it would be reasonable to refactor later.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -899,6 +899,14 @@ def __init__(self, controller: Any, title: str, *, zt_version: str,
         super().__init__(controller, widgets, 'ABOUT', popup_width, title)
 
 
+class MsgLinksView(PopUpView):
```
The related issue, #618, and the discussion [here](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/links) mentioned that we would want links to be listed at the end (with a different style) and combined with the `MsgInfoView`. 

On the contrary, this doesn't add links in `MsgInfoView` considering how 'Reactions' can potentially take a significant amount of space in the view and how it would be easier to work with a popup [than trailing links in `MessageBox`] later (e.g. handling internal/external links). Thoughts?

neiljp replied:
I'd suggest for a v1 we can integrate links as the end of a message, and also as a category above reactions in the message-info-view. I don't expect too many links in a single message?

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -200,11 +200,11 @@ def __init__(self, message: Message, model: Any,
                  last_message: Any) -> None:
         self.model = model
         self.message = message
+        self.message['links'] = OrderedDict()
```
Currently we treat the message data as essentially the data raw from the API. I'm a little concerned about adding extra data there since we don't know if the server may add a `links` data structure for messages, which this will then override.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -968,14 +968,20 @@ def __init__(self, controller: Any, color: str,
         ]
+        if message_links:
+            msg_info.append(('Message Links (see footlinks)', [
+                (str(index), text)
```
I attempted adding `text` and `link` but they disrupt `MsgInfoView` and trigger the wide column issue that we decided to tackle later in https://github.com/zulip/zulip-terminal/pull/659#discussion_r434257471. That said, even if we fix the issue, we'd end up with very wide `MsgInfoView` mostly.

I have added `index` and `text`, for now, but they don't seem to add much value.

Would it be desirable to introduce a message links popup inside `MsgInfoView` (like the edit history popup)?

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -476,14 +476,11 @@ def soup2markup(self, soup: Any, **state: Any) -> List[Any]:
-                else:
-                    markup.append(
-                        ('msg_link', '[' + text + ']' + '(' + link + ')'))
+                text = text if text else link
```
I was wondering if it would be better to trim the links that do not have any text to improve the flow of a message? We'd show the complete link in the footnote anyway.

preetmishra replied:
Yes, I meant the last segment. I have applied it in the latest update.

preetmishra commented:
@neiljp Thanks for the feedback. I have incorporated your suggestions in the latest update.

neiljp commented:
Another potentially straightforward change would be to simplify/compact any duplicate links, so that we don't end up with
```
1: google.com
2: duckduckgo.com
3: google.com
```
but rather just the first two and a re-use of [1] in the text in place of [3].

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -403,6 +403,23 @@ def reactions_view(self, reactions: List[Dict[str, Any]]) -> Any:
+            ('msg_footlink', indexed_link)
+            for indexed_link in indexed_links
+        ]
+        return urwid.Padding(urwid.Text(footlinks), align='left', left=10,
```
**Before**:

![Screenshot from 2020-07-06 23-11-05](https://user-images.githubusercontent.com/43616959/86623370-16d6a780-bfdf-11ea-881a-4048acef308d.png)

**After**:

![Screenshot from 2020-07-06 23-17-20](https://user-images.githubusercontent.com/43616959/86623386-1dfdb580-bfdf-11ea-878c-fd0854b9307f.png)

I prefer the _after_ as it adds a bit of distinction, between the content and the footlinks, and gives it more width. Additionally, I don't see much interference. Thoughts?

neiljp replied:
I'm not sure I have a strong preference at this stage, but I think this change will be easier to adjust later, and may change if we have either of
* link text & names (eg. link & google.com) styled as currently, but the numbers styled the same with the new style
* the ellipsis to avoid long links wrapping

preetmishra replied:
I agree. I have pushed the new update with all these three changes - namely _after_ alignment, new styling only for link indexes and ellipsis for now.

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -477,6 +478,7 @@ def soup2markup(self, soup: Any, **state: Any) -> List[Any]:
                     link = urljoin(self.model.server_url, link)
 
                 text = text if text else link
+                self.message_links.append((text, link))
```
I have amended and moved this in the commit where this is used and tested. The methods which use this change have the respective tests.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -812,13 +813,14 @@ def calculate_table_widths(contents: PopUpViewTableContent,
+                        len(max(text.split('\n'), key=len))
+                        for text in row
+                    ]
+                    strip_widths.append(max_row_lengths)
```
I haven't considered other widgets' width because they disrupt `MsgInfoView`.

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -476,14 +501,24 @@ def soup2markup(self, soup: Any, **state: Any) -> List[Any]:
                 else:
-                    markup.append(
-                        ('msg_link', '[' + text + ']' + '(' + link + ')'))
+                    saved_text, saved_link_index = self.message_links[link]
```
The else bit is to retain `text` if the link is already present in `message_links`.

preetmishra commented:
@neiljp Thanks for the comprehensive and thorough review.

I have reworked the commit structure. The first commit should only have the migration that is required from the previous view. Next, the subsequent commits should only have higher-level functionalities incrementally. The 4th and the 5th commit could be squashed together (I have split them for review).

Further, I have:
- Wrapped links in the footlinks with ellipses to save screen state.
- Handled duplicate links - https://github.com/zulip/zulip-terminal/pull/703#issuecomment-653816325.
- Styled link indexes differently than `link`/`text`.
- Trimmed link `text` to only show the last segment of the link if the text is redundant.
- Modified/hacked `PopUpView` to show message links in `MsgInfoView` properly.
- Used the suggested style in https://github.com/zulip/zulip-terminal/pull/703#pullrequestreview-442628731 for `MsgInfoView`.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -402,6 +405,28 @@ def reactions_view(self, reactions: List[Dict[str, Any]]) -> Any:
             return ''
 
+    # Use quotes as a workaround for OrderedDict typing issue.
+    # See https://github.com/python/mypy/issues/6904.
```
Thanks for the comment :+1:

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -402,6 +405,28 @@ def reactions_view(self, reactions: List[Dict[str, Any]]) -> Any:
+    # See https://github.com/python/mypy/issues/6904.
+    def footlinks_view(self,
+                       message_links: 'OrderedDict[str, Tuple[str, int]]',
+                       ) -> Any:
```
While we don't currently type return type of urwid widgets, I think it might be clearer if we return `None` instead of the empty string, so that we are effectively returning `Optional[Any]`. That's not worth annotating now, but if we do move to a stage of urwid widget types, `Union[str, <widget>]` will be less clear than `Union[None, <widget>]`, ie. `Optional[<widget>]`.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -476,14 +501,24 @@ def soup2markup(self, soup: Any, **state: Any) -> List[Any]:
-                    markup.append(
-                        ('msg_link', '[' + text + ']' + '(' + link + ')'))
+                    saved_text, saved_link_index = self.message_links[link]
+                    if saved_text != text:
+                        self.message_links[link] = (
+                            '{}, {}'.format(saved_text, text), saved_link_index
+                        )
```
You don't have a test for this?

preetmishra replied:
The 'saved text' is not used anywhere currently and is for the popup (where we'll show `1: Text \n Link`). However, the code is still useful to have so that we don't miss the 'saved text' even though we're not using it.

I have added extra test cases that are remotely related.

neiljp commented on zulipterminal/config/themes.py:
```diff
@@ -29,6 +29,7 @@
     'reaction_mine': 'standout',
     'msg_mention': 'bold',
     'msg_link': '',
+    'msg_link_index': 'bold',
```
While the previous commit is fine first, if you have this change first, you don't need to change all the uses of it to be the new names.

preetmishra replied:
True. Notwithstanding, the changes in this particular commit express where the new style is used explicitly [in GitHub's diff].

neiljp replied:
I agree that the ordering you have is reasonable, it just ends up with a larger diff.

preetmishra replied:
Yeah, that's there. However, it only fits with the previous commit which is pretty large already.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1526,21 +1526,18 @@ def test_private_message_to_self(self, mocker):
           ('msg_link', 'Bar'), ' ', ('msg_link_index', '[2]')]),
         ('<a href="http://foo">http://foo</a>',
-         [('msg_link', 'http://foo'), ' ', ('msg_link_index', '[1]')]),
+         [('msg_link', 'foo'), ' ', ('msg_link_index', '[1]')]),
```
This seems unexpected, but it might just be because this 'domain name' is so short.

This is something of a tricky problem, as I've seen links which are eg. query strings, so the text becomes `some_file?this_thing`, which looks unexpected (and real cases even more so). I'm reconsidering how we should handle them, though I don't see an easy solution straight away.

For now this looks OK.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -502,6 +502,9 @@ def soup2markup(self, soup: Any, **state: Any) -> List[Any]:
                 text = text if text else link
+                # Only use the last segment if the text is a link itself.
+                if text == link:
+                    text = text.rstrip('/').split('/')[-1]
```
We could consider parsing the URL using the standard library, as we do in other places.

preetmishra replied:
Resolved in PM. (We're moving ahead without the library call for now due to few inconsistencies. Though, we may want to switch to it later.)

neiljp commented:
@preetmishra I'm using this now, and I actually wonder if we should avoid showing links in the footlinks if they're just eg. `example.com` with no text - we don't gain anything? The user doesn't gain any information other than perhaps continuity of numbering - seeing example.com twice is actually more text than we had before! It also exposes how we're implementing the auto link/text association, after a message author just wrote what looks like a link with no description. They could still be shown in other contexts (popups).

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1654,6 +1669,7 @@ def test_soup2markup(self, content, markup):
         self.model.server_url = SERVER_URL
         # NOTE Absence of previous (last) message should not affect markup
         msg_box = MessageBox(message, self.model, None)
+        msg_box.link_index = 1
```
It was necessary as I was initializing the variable in `soup2markup`'s caller method. That said, I gather it better belongs to the `__init__`.

Re `len(message_links)`, I am not sure if this would work here.

neiljp replied:
By my reading, `link_index` is really 'next link index', so equivalent to `len(message_links)+1`? So we could avoid needing to synchronize two variables.

preetmishra replied:
Thanks for the tip! I have implemented it in the latest update.

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -468,22 +494,33 @@ def soup2markup(self, soup: Any, **state: Any) -> List[Any]:
-                link = element.attrs['href']
+                link = element.attrs['href'].rstrip('/')
                 text = element.img['src'] if element.img else element.text
+                text = text.rstrip('/')
```
The rstrips are useful to have to avoid anomalies and edge cases like https://google.com vs https://google.com/.

neiljp replied:
This would be a useful comment :)

preetmishra commented:
@neiljp Thanks for the review. I have incorporated your suggestions in the latest update and replied to your other in-line comments. Moreover, I have introduced a new commit to address https://github.com/zulip/zulip-terminal/pull/703#issuecomment-656915338.

Additionally, I have dropped the popup commits for now. I will push a new PR after #695 is settled.

preetmishra commented:
Updated with the comment in https://github.com/zulip/zulip-terminal/pull/703#discussion_r453336979.

preetmishra commented:
Strange! I pushed the commits successfully when GitHub was down, but they were not showing up. I had to `--amend --no-edit` my commit and force push again.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -402,6 +405,28 @@ def reactions_view(self, reactions: List[Dict[str, Any]]) -> Any:
+            ])
+        footlinks[-1] = footlinks[-1][:-1]  # Remove the last newline.
+
+        return urwid.Padding(urwid.Text(footlinks, wrap='ellipsis'),
```
As we discussed elsewhere, when we add the ellipsis, we should add a FAQ entry regarding this.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -413,6 +413,10 @@ def footlinks_view(self,
                        ) -> Any:
         footlinks = []
         for link, (text, index) in message_links.items():
+            # Skip the footlink if its text only has the network location.
+            if text == urlparse(link).netloc:
+                continue
```
I'm not sure if it's just this code/commit or the combination with previous commits - this works well for simple cases that I tested manually, but I think we may wish to be careful with handling the link scheme...or even the entire link?

Given a message of `example.com http://example.com https://example.com`, the webapp shows 3 links expressed as they are in the message (with 'text' exactly as written in the message), the first two pointing to the same address. We point the first two to the same address (based on the HTML, presumably), but we remove the scheme, so they *look* identical but are confusingly not, ie. we generate `example.com [1] example.com [1] example.com [2]`. Given we don't show the footlinks for these now after this commit, this is rather confusing! So, at a minimum, I think we should leave the scheme as it is, and probably use the assumed href in the anchor where no scheme is present?

After some further manual testing, note that even a long link like eg `example.com/blah/foooo/agikhasdglkashdglkashdg/asdfgklhasg/asdjkgalskjdhalskjdhlkasdjhsalkdghsasiougypoadisgypsdgypodsgypoadgysapdgy` gets shown twice identically (in message and footlinks). I don't think we gain from this being in the footlinks, much like with the simple `example.com` case? So, if there is no text, this makes me think we should remove it in all cases. Thoughts?

preetmishra replied:
Thanks for the report. It turns out that the server translates any scheme-less link to `HTTP` (e.g. `example.com` becomes `<a href="http://example.com">example.com</a>`). The core problem was with the `text == link` check that I had previously. I have replaced it with a 'scheme-less' check to figure out whether the 'text' part is identical.

Moreover, I have amended the last segment bit to only work when the text is not too simple (e.g. `example.com`) and to patch texts like these with their link itself. Consequently, `example.com http://example.com https://example.com` come out as `http://example.com [1] http://example.com [1] https://example.com [2]` and long links work more reliably.

Furthermore, I have not redacted the last segment bit for long links as I believe they improve the reading flow drastically. We could surely improve how we detect the 'significant' segment later but I think they work well for now.

preetmishra replied:
(See also https://github.com/zulip/zulip-terminal/pull/703/commits/a782d546470df674c78e7d52268fd89765b405f0#diff-3e518af2a88ffb14e1771a621641edd5R1552-R1558.)

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1515,17 +1516,35 @@ def test_private_message_to_self(self, mocker):
+          ('msg_link', 'Bar'), ' ', '[2]']),
+        ('<a href="http://foo">Foo</a><a href="http://foo">Another foo</a>',
+         [('msg_link', 'Foo'), ' ', '[1]',
+          ('msg_link', 'Another foo'), ' ', '[1]']),
```
minor: A variant of this showing the stripping of trailing slashes (maybe even with identical text?) would be useful, as I don't think we have a test of that?

neiljp commented:
Regarding the hiding of all full-text links, I should clarify that if there is a 'file name' then this works well as the 'text' substitute, but otherwise I'm not sure that another part of the URL makes sense on its own.

preetmishra commented:
@neiljp Thanks for the review and discovering those examples. They helped me greatly to improve the last segment and the skip footlink bit. :+1: 

I have removed the `link_index` attribute and moved to only using the `message_links`, and added tests for the trailing slash bit. Moreover, I have reworked the last segment detection bit and the related skip footlink. Hopefully, they should be more reliable now.

Thanks for helping me with the commit structure. They do seem to tell a 'story' with the incremental changes and look coherent when compared to what we started with. :+1: The duplicate links excerpt is necessary to have with the initial commit otherwise the `OrderedDict` keeps getting overridden and we end up with rather funny link indexes.

Re hiding of all full-text links, see https://github.com/zulip/zulip-terminal/pull/703#discussion_r454576011.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1536,19 +1536,26 @@ def test_private_message_to_self(self, mocker):
+         '<a href="http://foo.com">http://foo.com</a>'
+         '<a href="https://foo.com">https://foo.com</a>'
+         '<a href="http://foo.com">Text</a>',
+         [('msg_link', 'http://foo.com'), ' ', ('msg_link_index', '[1]'),
```
This is my only remaining query. It's minor really, but if you look at #**commits** then you'll see some on a `chat.zulip.org` branch - it gets shown as `http://chat.zulip.org`, which of course isn't the branch name :)

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -412,6 +412,11 @@ def footlinks_view(self,
                        ) -> Any:
         footlinks = []
         for link, (text, index) in message_links.items():
+            # Skip the footlink if it is identical to its text (see soup2mark
```
`soup2markup`?

preetmishra commented:
@neiljp Thanks for the review! In the latest update, I have:
* Fixed the link text behaviour that you mentioned here and we talked about on CZO. Now, `example.com http://example.com https://example.com` come out as `example.com [1] http://example.com [1] https://example.com [2]` (without any footlink) and the last segment is only used when the link text has something else than just the 'domain name'.
* Improved the tests and added the mentioned TODO comment.
* Restructured the tests to amend them incrementally as the refinements are added.

neiljp commented:
@preetmishra Thanks for the update and all your work on this :+1: I've just merged this! :tada:

The only outstanding thing I think we could tidy up is the cases where we show the last entry (or not), but that's definitely fine tuning now, and we have a lot of test cases that we can re-examine to determine if we want different behavior. On that note, it might be worth investigating if we can incorporate the ids inline (`pytest.param`?), since we have so many parametrizations in `soup2markup` now!

The popup and link-opening work will be great to build off this. Perhaps we can use the same parametrizations as a fixture?

preetmishra commented:
@neiljp Thanks for the merge! I'll proceed with the follow-ups now. :+1:
